// power-assign/dist/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.normalizeOperation = exports.toJSON = exports.mergeOperations = exports.retargetToPropWithRestoration = exports.retargetToProp = exports.assignToPropWithRestoration = exports.assignWithRestoration = exports.assignToProp = exports.assign = undefined;

var _powerAssign = require('./power-assign.js');

Object.defineProperty(exports, 'assign', {
  enumerable: true,
  get: function get() {
    return _powerAssign.assign;
  }
});
Object.defineProperty(exports, 'assignToProp', {
  enumerable: true,
  get: function get() {
    return _powerAssign.assignToProp;
  }
});
Object.defineProperty(exports, 'assignWithRestoration', {
  enumerable: true,
  get: function get() {
    return _powerAssign.assignWithRestoration;
  }
});
Object.defineProperty(exports, 'assignToPropWithRestoration', {
  enumerable: true,
  get: function get() {
    return _powerAssign.assignToPropWithRestoration;
  }
});

var _retargetToProp = require('./retarget-to-prop.js');

Object.defineProperty(exports, 'retargetToProp', {
  enumerable: true,
  get: function get() {
    return _retargetToProp.retargetToProp;
  }
});
Object.defineProperty(exports, 'retargetToPropWithRestoration', {
  enumerable: true,
  get: function get() {
    return _retargetToProp.retargetToPropWithRestoration;
  }
});

var _oadUtils = require('oad-utils');

Object.defineProperty(exports, 'mergeOperations', {
  enumerable: true,
  get: function get() {
    return _oadUtils.mergeUpdateOperations;
  }
});
Object.defineProperty(exports, 'toJSON', {
  enumerable: true,
  get: function get() {
    return _oadUtils.updateOperationToJSON;
  }
});
Object.defineProperty(exports, 'normalizeOperation', {
  enumerable: true,
  get: function get() {
    return _oadUtils.normalizeUpdateOperation;
  }
});

var _powerAssign2 = _interopRequireDefault(_powerAssign);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = _powerAssign2.default;
